File: practice/algorithms/dynamic_programming/angry_children/angry_children.py
#!/usr/bin/env python

n = int(raw_input().strip())
k = int(raw_input().strip())
candies = []
for _ in range(n):
    candies.append(int(raw_input().strip()))

candies.sort()

ans = 10 ** 20
s = (k - 1) * candies[k - 1]
for POS in range(k - 1):
    s += (2 * POS + 1 - k) * candies[POS]

part_sums = [0]
for POS in range(n):
    part_sums.append(part_sums[-1] + candies[POS])
    
for POS in range(n - k):
    if s < ans:
        ans = s
    s += (k - 1) * (candies[POS + k] + candies[POS])
    s -= 2 * (part_sums[k + POS] - part_sums[POS + 1])
    
if s < ans:
    ans = s
print ans

File: practice/algorithms/dynamic_programming/angry_children/angry_children.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <limits>

using namespace std;

int main()
{
    int n, k;
    cin >> n >> k;
    
    vector<long long int> candies(n);
    for (int pos = 0; pos < n; ++pos)
    {
        cin >> candies[pos];
    }
    
    sort(candies.begin(), candies.end());
    
    long long int s = (k - 1) * candies[k - 1];
    for (int pos = 0; pos < k - 1; ++pos)
    {
        s += (2 * pos + 1 - k) * candies[pos];
    }
    
    
    vector<long long int> psums(n + 1);
    for (int pos = 0; pos < n; ++pos)
    {
        psums[pos + 1] = psums[pos] + candies[pos];
    }
    
    long long int ans = numeric_limits<long long int>::max();
    for (int pos = 0; pos < n - k; ++pos)
    {
        if (s < ans)
        {
            ans = s;
        }
        s += (k - 1) * (candies[pos + k] + candies[pos]);
        s -= 2 * (psums[pos + k] - psums[pos + 1]);
    }
    
    if (s < ans)
    {
        ans = s;
    }
    
    cout << ans << endl;
    
    return 0;
}

